File: java/PeekingIterator.java
package practice01;

import java.util.Iterator;
import java.util.ListIterator;

public class PeekingIterator implements Iterator<Integer> {

	private boolean flag;
	private Integer integer;
	private Iterator<Integer> iterator;
	public PeekingIterator(Iterator<Integer> iterator) {
	    // initialize any member here.
	    this.iterator = iterator; 
	    flag = false;
	}

    // Returns the next element in the iteration without advancing the iterator.
	public Integer peek() {
        if (!flag) {
        	flag = true;
        	integer = iterator.next();
        }
        return integer;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	@Override
	public Integer next() {
	    if (!flag) return iterator.next();
	    flag = false;
	    return integer;
	}

	@Override
	public boolean hasNext() {
	    if (!flag) return iterator.hasNext();
	    return true;
	}

}

File: python/PeekingIterator.py
class PeekingIterator(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.iterator = iterator
        self.peeked = False
        self.nextElem = None

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if self.peeked == False:
            self.peeked = True
            self.nextElem = self.iterator.next()
        return self.nextElem

    def next(self):
        """
        :rtype: int
        """
        if self.peeked:
            self.peeked = False
            nextElem = self.nextElem
            self.nextElem = None
            return nextElem
        return self.iterator.next()

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.peeked or self.iterator.hasNext()